File: pie_sol/pie_sol/utils/py/warranty.py
import frappe

def date_validate(doc, events):
    previous_actual_date = None  # Initialize previous actual service date outside the loop

    for i, row in enumerate(doc.table_hsoa):
        current_date = row.get('actual_service_date')   
        print(current_date,'@@@@@@@@@@@@@@@@@@@@@@@')
        print(previous_actual_date,'$$$$$$$$$$$$$$$$$$$$$$$$$$')
        if current_date and previous_actual_date:
            if previous_actual_date > current_date:
                frappe.throw(f"Date in row {i+1} cannot be less than the date in the previous row")
        
        previous_actual_date = current_date  # Update previous actual service date for next iteration
        print(previous_actual_date,'&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&&')



def machine_id_verify(doc,events):

    temp=frappe.db.get_list("Warranty",{'machine':doc.machine})
    if len(temp)>=1:
        
        frappe.throw("Machine Id already linked with another Warranty")



@frappe.whitelist()
def create_warranty(customer_name,machine_name,warranty_period_in_month,warranty_service_count,warranty_start_date,warranty_contract_date,warranty_serviced_by,Warranty_enrolment):
    print('*********************')
    warranty = frappe.new_doc("Warranty")
    warranty.customer_name=customer_name
    warranty.machine = machine_name
    warranty.service_count = warranty_service_count
    warranty.period_in_month = warranty_period_in_month
    warranty.start_date = warranty_start_date
    warranty.contract_date = warranty_contract_date
    warranty.serviced_by = warranty_serviced_by
    warranty.enrolment = Warranty_enrolment
    # Set more fields here as needed
    warranty.save()

File: pie_sol/pie_sol/utils/js/warranty.js

// frappe.ui.form.on("Warranty", {
//     refresh(frm) {
//         cur_frm.fields_dict['table_hsoa'].$wrapper.find(".grid-add-row")[0].style.display="none";
        
//         frm.set_query("machine", function () {
//             return {
//               filters: {
//                 customer_name:frm.doc.customer_name
//               },
//             };
            
//           }
//         );
//     },

//     period_in_month:function(frm) {
//         console.log("1")
//         if(frm.is_new()) {
//             var n = frm.doc.service_count
//             start_date=frm.doc.start_date
//             period_in_month=frm.doc.period_in_month
//             service_count=frm.doc.service_count
//             start_date=frappe.datetime.add_months(start_date,months=period_in_month);
//             temp=""
//             for (var i = 0; i < n; i++) {

//                 var child = frm.add_child('table_hsoa');
//                 console.log("786")
//                 child.planned_service_date = frappe.datetime.add_months(start_date,months=i*period_in_month);
//                 temp=child.planned_service_date
//                 console.log(temp)
//             }
//             console.log(temp)
//             frm.doc.end_date=temp;
//             console.log(frm.doc.end_date)
//         frm.refresh_field('table_hsoa');
//         frm.refresh_field('end_date');
//         }
//     },




// });
